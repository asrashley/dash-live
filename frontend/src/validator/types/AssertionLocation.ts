export interface AssertionLocation {
    filename: string;
    line: number;
}
