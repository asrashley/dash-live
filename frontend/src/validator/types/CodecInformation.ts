export interface CodecDetails {
    label: string;
    error?: string;
    details: string[];
}

export interface CodecInformation {
    codec: string;
    error?: string;
    details: CodecDetails[];
}
