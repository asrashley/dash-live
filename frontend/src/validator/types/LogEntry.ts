export type LogEntry = {
    id: number;
    level: 'error' | 'info' | 'debug';
    text: string;
};
