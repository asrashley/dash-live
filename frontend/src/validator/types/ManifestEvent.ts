
export interface ManifestEvent {
    text: string[];
}