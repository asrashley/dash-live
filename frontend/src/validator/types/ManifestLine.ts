export interface ManifestLine {
    line: number;
    text: string;
    hasError: boolean;
    errors: string[];
}
