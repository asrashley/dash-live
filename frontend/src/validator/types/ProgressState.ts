export interface ProgressState {
    currentValue?: number;
    minValue: number;
    maxValue: number;
    text: string;
    finished?: boolean;
    error?: boolean;
 }
