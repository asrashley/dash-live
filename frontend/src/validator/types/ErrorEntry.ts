import { AssertionLocation } from "./AssertionLocation";

export interface ErrorEntry {
    assertion: AssertionLocation;
    location: [number, number];
    clause?: string;
    msg: string;
}