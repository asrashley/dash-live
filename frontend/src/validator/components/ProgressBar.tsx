import { useComputed, type ReadonlySignal } from "@preact/signals";
import { ProgressState } from "../types/ProgressState";

export interface ProgressBarProps {
  progress: ReadonlySignal<ProgressState>;
}

export function ProgressBar({ progress }: ProgressBarProps) {
  const className = useComputed(
    () => {
      const pg = progress.value;
      let cls = "progress-bar progress-bar-striped";
      if (pg.currentValue === undefined) {
        cls += " visually-hidden";
      } else if (pg.finished) {
        cls += pg.error ? " bg-warning text-dark" : " bg-success";
      } else {
        cls  += " bg-info progress-bar-animated";
      }
      return cls;
    }
  );
  const pct = useComputed<string>(
    () => {
      const { minValue, maxValue, currentValue } = progress.value;
      return `${(100 * (currentValue - minValue)) / Math.max(1, maxValue - minValue)}%`;
    });
  const style = useComputed(() => ({
    width: pct.value,
  }));

  return (
    <div>
      <div
        className={className}
        style={style}
        role="progressbar"
        aria-valuenow={progress.value.currentValue}
        aria-valuemin={progress.value.minValue}
        aria-valuemax={progress.value.maxValue}>
        {pct}
      </div>
      <div className="progress-text">{progress.value.text}</div>
    </div>
  );
}
