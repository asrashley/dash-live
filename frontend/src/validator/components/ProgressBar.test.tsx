import { describe, expect, test } from "vitest";
import { signal } from "@preact/signals";

import { renderWithProviders } from "../../test/renderWithProviders";
import { ProgressBar } from "./ProgressBar";
import { ProgressState } from "../types/ProgressState";

describe("ProgressBar component", () => {
  const progress = signal<ProgressState>();

  test.each([
    [undefined, undefined],
    [0, undefined],
    [10, "started"],
    [25, "getting busy"],
    [50, "half way there"],
    [75, "not long now"],
    [100, "all done"],
  ])("matches snapshot for %s%%", (pct: number | undefined, txt: string | undefined) => {
    progress.value = {
      text: txt,
      currentValue: pct,
      minValue: 0,
      maxValue: 100,
      finished: pct === 100,
    };
    const { asFragment, getBySelector, getByText } = renderWithProviders(
      <ProgressBar progress={progress} />
    );
    if (txt) {
        getByText(txt);
    }
    const bar = getBySelector(".progress-bar") as HTMLElement;
    if (pct !== undefined) {
        expect(bar.style.width).toEqual(`${pct}%`);
    }
    expect(bar.classList.contains("bg-success")).toEqual(pct === 100);
    expect(bar.classList.contains('bg-warning')).toEqual(false);
    expect(bar.classList.contains("progress-bar-animated")).toEqual(pct !== undefined && pct !== 100);
    expect(bar.classList.contains("visually-hidden")).toEqual(pct === undefined);
    expect(asFragment()).toMatchSnapshot();
  });

  test('finished with an error', () => {
    progress.value = {
      text: '',
      currentValue: 34.5,
      minValue: 0,
      maxValue: 100,
      finished: true,
      error: true,
    };
    const { getBySelector } = renderWithProviders(
      <ProgressBar progress={progress} />
    );
    const bar = getBySelector(".progress-bar") as HTMLElement;
    expect(bar.classList.contains('bg-warning')).toEqual(true);
    expect(bar.classList.contains('bg-success')).toEqual(false);
  });
});
