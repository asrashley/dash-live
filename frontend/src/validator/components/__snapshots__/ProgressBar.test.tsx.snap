// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgressBar component > matches snapshot for 0% undefined 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      class="progress-bar progress-bar-striped bg-info progress-bar-animated"
      role="progressbar"
      style="width: 0%;"
    >
      0%
    </div>
    <div
      class="progress-text"
    />
  </div>
</DocumentFragment>
`;

exports[`ProgressBar component > matches snapshot for 10% started 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="10"
      class="progress-bar progress-bar-striped bg-info progress-bar-animated"
      role="progressbar"
      style="width: 10%;"
    >
      10%
    </div>
    <div
      class="progress-text"
    >
      started
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBar component > matches snapshot for 25% getting busy 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="25"
      class="progress-bar progress-bar-striped bg-info progress-bar-animated"
      role="progressbar"
      style="width: 25%;"
    >
      25%
    </div>
    <div
      class="progress-text"
    >
      getting busy
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBar component > matches snapshot for 50% half way there 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="50"
      class="progress-bar progress-bar-striped bg-info progress-bar-animated"
      role="progressbar"
      style="width: 50%;"
    >
      50%
    </div>
    <div
      class="progress-text"
    >
      half way there
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBar component > matches snapshot for 75% not long now 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="75"
      class="progress-bar progress-bar-striped bg-info progress-bar-animated"
      role="progressbar"
      style="width: 75%;"
    >
      75%
    </div>
    <div
      class="progress-text"
    >
      not long now
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBar component > matches snapshot for 100% all done 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="100"
      class="progress-bar progress-bar-striped bg-success"
      role="progressbar"
      style="width: 100%;"
    >
      100%
    </div>
    <div
      class="progress-text"
    >
      all done
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProgressBar component > matches snapshot for undefined% undefined 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      class="progress-bar progress-bar-striped visually-hidden"
      role="progressbar"
    >
      NaN%
    </div>
    <div
      class="progress-text"
    />
  </div>
</DocumentFragment>
`;
